sdk/opa.go

```diff
@@ -0,0 +1,238 @@
+	configBytes            []byte
+	paths                  []string
+	filter                 loader.Filter
+	bundleMode             bool
+	store                  storage.Store
+	preparedQueriesDoOnce  map[string]*sync.Once
+	preparedQueries        map[string]*rego.PreparedEvalQuery
+	interQueryBuiltinCache iCache.InterQueryCache
+}
+
+// ConfigFile sets the configuration file to use on the OPA instance.
+func ConfigFile(fileName string) func(*OPA) error {
+	return func(opa *OPA) error {
+		bs, err := ioutil.ReadFile(fileName)
+		if err != nil {
+			return err
+		}
+		opa.configBytes = bs
+		return nil
+	}
+}
+
+// Config sets the configuration to use on the OPA instance.
+func Config(config string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.configBytes = []byte(config)
+		return nil
+	}
+}
+
+// Paths sets the paths to search for policy, data files and bundles
+func Paths(paths []string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.paths = paths
+		return nil
+	}
+}
+
+// Store allows providing storage implementation for OPA. Will use inmem.New() if not provided.
+func Store(store storage.Store) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.store = store
+		return nil
+	}
+}
+
+// New returns a new OPA object.
+func New(opts ...func(*OPA) error) (*OPA, error) {
+	opa := &OPA{}
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginInitializers []func(*plugins.Manager)
+
+	if opa.paths != nil && len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, fmt.Errorf("load error %w", err)
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	if opa.store == nil {
+		opa.store = inmem.New()
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, opa.store, pluginInitializers...)
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.RegisterCompilerTrigger(opa.compilerUpdated)
+	opa.manager.Register("discovery", discovery)
+
+	opa.interQueryBuiltinCache = iCache.NewInterQueryCache(opa.manager.InterQueryBuiltinCacheConfig())
+
+	opa.preparedQueries = map[string]*rego.PreparedEvalQuery{}
+	opa.preparedQueriesDoOnce = map[string]*sync.Once{}
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (interface{}, error) {
```

Comment:
Good idea. I've now added a `QueryDefault` which does this.

Reply:
Sorry to nitpick this...what about flipping it around so that `Query()` calls the default decision and we rename this one to something like `NamedQuery`? My reasoning is that the simplest example will want to omit the decision name.